app/src/main/java/com/ikarabulut/RuleSet.java

```diff
@@ -0,0 +1,12 @@
+package com.ikarabulut;
+
+public interface RuleSet {
+
+    public boolean willBeAliveNextGeneration(int aliveNeighborsCount, boolean isAlive);
+
+    private boolean isUnderpopulated() { return false; }
```

Comment:
Normally when defining an interface, the method signatures are given, but there is no body like in `willBeAliveNextGeneration` on line 5.

Since you provided default implementations in the interface, what happens if you remove the private methods from one of the Rules?

Reply:
Also, when defining an interface we rarely, if ever, define private methods.  Interfaces should describe what the _public_ contract of an object is, not describe its implementation details. The consumer of an interface like List or RuleSet does not need to know how the abstraction is implemented. Only what methods can I call on you, and what will they return.

Reply:
tbh, default method implementations in interfaces is a relatively new feature in Java.  I'm now genuinely curious what happens when you remove the private methods from a Rules object.  Do they inherit the default implementations defined in the interface?  If you change the signature of one of the private methods, does it cause a type error in any of the callers to require it to be updated?

Reply:
I set them to be default instead. So the implementation details are still technically public but they are only housed in the interface and not typed out in the actual class

Reply:
So here is the trade-off we're making:

Benefit: we saved some code duplication.
Cost: we polluted the public API of RuleSet with methods none of the callers want.

I asked for interfaces here to teach this concept, so you didn't make a mistake.  But, now is a good time to think about what would the trade-offs be here if we used Inheritance to achieve polymorphism instead?

(I'm not advocating for object inheritance as "better".  But, I think in this specific case, the trade-offs are interesting.)

Reply:
> we polluted the public API of RuleSet with methods none of the callers want.

This is a perfect time to introduce the "I" in SOLID.  The Interface Segregation Principle. :)  Not quite exactly what we're talking about, but related to the point I'm making about polluting a public API with methods callers don't need.

https://en.wikipedia.org/wiki/Interface_segregation_principle